### ISSUE TITLE
Compiler crash: "`` at  has no type (Exception)"

### ISSUE THREAD

**Issue by anykeyh** [crystal-lang/crystal#7304]
## Behavior

Compiler is not happy with the code below and is going to crash.

## Notes

According to the error message, it may be duplicate of #7017, but I don't think so.

## Example code 

```crystal
class MyClass
  @@connections = {} of String => Channel(String)

  alias Key = {String, Fiber}
  alias Value = { String, Int32 }
  @@fiber_connections : Hash(Key, Value) = {} of Key => Value

  def self.test(target : String, &block)
    fiber_target = {target, Fiber.current}
    channel = @@connections.fetch(target){ raise "UNFOUND" }
    db, call_count = @@fiber_connections.fetch(fiber_target){ {channel.receive, 0} }
  end

end

MyClass.test("default"){}
```

## Current work-around

Strangely, initial call to hash setter will let the compiler compile. This code below COMPILE

```crystal
class MyClass
  @@connections = {} of String => Channel(String)

  alias Key = {String, Fiber}
  alias Value = { String, Int32 }
  @@fiber_connections : Hash(Key, Value) = {} of Key => Value

  def self.init
    @@connections["t"] = Channel(String).new #< Any data in @@connections and it works!
  end

  def self.test(target : String, &block)
    fiber_target = {target, Fiber.current}
    channel = @@connections.fetch(target){ raise "UNFOUND" }
    db, call_count = @@fiber_connections.fetch(fiber_target){ {channel.receive, 0} }
  end

end

MyClass.init
MyClass.test("default"){}
```

## Additional notes

- Related to https://github.com/anykeyh/clear/issues/61
- Real case where it happens can be found here: https://github.com/anykeyh/clear/blob/master/src/clear/sql/connection_pool.cr
- The problem is no big deal in this real-case, since the "work-around" is the normal expected behavior.

**Comment by anykeyh** [comment]
Changing `Channel` to `Channel::Buffered` removed the crash.  I tried to play around with abstract, generic, abstract and generic etc.. but I can't reproduce unless with `Channel` as object :unamused:

**Comment by jhass** [comment]
It looks like this was fixed in 0.31 https://carc.in/#/r/8kw1
